key in command for ExecutionFailedException

I am using Kyro and trying to extend a the log and default recovery policy to delete the key that has a serialization issue.

When looking at the Reports trait, the message function says I get a ExecutionFailedException instead of the expected SerializationException i see later in the report stack trace.

Also, the key value seems to be the command and the command has "<command> <key>"seems to be appearing as part of the command and the command :
Command $command for key '$key' failed -> Command GET <MyKeyName> for key 'GET' failed.

```
17:52:29 ERROR play.api.cache.redis - Command GET hasFlags.17592186045482 for key 'GET' failed.
play.api.cache.redis.exception.SerializationException: Deserialization failed for hasFlags.17592186045482
        at play.api.cache.redis.exception.package$.serializationFailed(package.scala:18)
        at play.api.cache.redis.connector.RedisConnectorImpl$$anonfun$play$api$cache$redis$connector$RedisConnectorImpl$$decode$1.applyOrElse(RedisConnectorImpl.scala:73)
        at play.api.cache.redis.connector.RedisConnectorImpl$$anonfun$play$api$cache$redis$connector$RedisConnectorImpl$$decode$1.applyOrElse(RedisConnectorImpl.scala:72)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:34)
        at scala.util.Failure.recover(Try.scala:230)
        at play.api.cache.redis.connector.RedisConnectorImpl.play$api$cache$redis$connector$RedisConnectorImpl$$decode(RedisConnectorImpl.scala:72)
        at play.api.cache.redis.connector.RedisConnectorImpl$$anonfun$get$1.applyOrElse(RedisConnectorImpl.scala:51)
        at play.api.cache.redis.connector.RedisConnectorImpl$$anonfun$get$1.applyOrElse(RedisConnectorImpl.scala:48)
        at scala.PartialFunction$OrElse.apply(PartialFunction.scala:168)
        at scala.util.Success.$anonfun$map$1(Try.scala:251)
        at scala.util.Success.map(Try.scala:209)
        at scala.concurrent.Future.$anonfun$map$1(Future.scala:287)
        at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
        at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:38)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
        at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: com.esotericsoftware.kryo.KryoException: Encountered unregistered class ID: 2013895718
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:137)
        at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:693)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:804)
        at com.romix.akka.serialization.kryo.KryoBasedSerializer.fromBinary(KryoSerializer.scala:504)
        at com.romix.akka.serialization.kryo.KryoSerializer.fromBinary(KryoSerializer.scala:351)
        at akka.serialization.Serialization.$anonfun$deserialize$3(Serialization.scala:203)
        at scala.util.Try$.apply(Try.scala:209)
        at akka.serialization.Serialization.deserialize(Serialization.scala:203)
        at play.api.cache.redis.connector.AkkaDecoder.binaryToAnyRef(AkkaSerializer.scala:121)
        at play.api.cache.redis.connector.AkkaDecoder.$anonfun$stringToAnyRef$2(AkkaSerializer.scala:125)
        at scala.Function1.$anonfun$andThen$1(Function1.scala:52)
        at play.api.cache.redis.connector.AkkaDecoder.stringToAnyRef(AkkaSerializer.scala:125)
        at play.api.cache.redis.connector.AkkaDecoder.untypedDecode(AkkaSerializer.scala:112)
        at play.api.cache.redis.connector.AkkaDecoder.decode(AkkaSerializer.scala:96)
        at play.api.cache.redis.connector.AkkaSerializerImpl.$anonfun$decode$1(AkkaSerializer.scala:162)
        at scala.util.Try$.apply(Try.scala:209)
        at play.api.cache.redis.connector.AkkaSerializerImpl.decode(AkkaSerializer.scala:162)
        ... 21 common frames omitted
```

--- a/README.md
+++ b/README.md
@@ -415,6 +415,11 @@ both implementations along.
 
 ## Changelog
 
+### [:link: 1.5.1](https://github.com/KarelCemus/play-redis/tree/1.5.1)
+
+Fixed [#102](https://github.com/KarelCemus/play-redis/issues/102), preserved original 
+exception if extends `RedisException` and fixed wrong parameters in error messages
+
 ### [:link: 1.5.0](https://github.com/KarelCemus/play-redis/tree/1.5.0)
 
 **[Scredis](https://github.com/scredis/scredis) connector replaced 
@@ -432,6 +437,11 @@ Changed return type `Unit` in `AbstractCacheApi` to `akka.Done` with the same me
 but better signalizing the intention.
 
 Cross-compiled for Scala 2.11 and Scala 2.12.
+
+### [:link: 1.4.2](https://github.com/KarelCemus/play-redis/tree/1.4.2)
+
+Fixed [#102](https://github.com/KarelCemus/play-redis/issues/102), preserved original 
+exception if extends `RedisException` and fixed wrong parameters in error messages
 
 ### [:link: 1.4.1](https://github.com/KarelCemus/play-redis/tree/1.4.1)
 

--- a/src/main/scala/play/api/cache/redis/connector/ExpectedFuture.scala
+++ b/src/main/scala/play/api/cache/redis/connector/ExpectedFuture.scala
@@ -37,7 +37,5 @@ private[ connector ] class ExpectedFuture[ T ]( future: Future[ T ], key: String
   */
 private[ connector ] class ExpectedFutureBuilder[ T ]( future: Future[ T ] ) {
 
-  def executing( key: String, cmd: => String ): ExpectedFuture[ T ] = new ExpectedFuture[ T ]( future, key, cmd )
-
   def executing( key: String ): ExpectedFuture[ T ] = new ExpectedFuture[ T ]( future, key, key )
 }

--- a/src/main/scala/play/api/cache/redis/connector/ExpectedFuture.scala
+++ b/src/main/scala/play/api/cache/redis/connector/ExpectedFuture.scala
@@ -10,32 +10,64 @@ import play.api.cache.redis.exception._
   *
   * @author Karel Cemus
   */
-private[ connector ] class ExpectedFuture[ T ]( future: Future[ T ], key: String, cmd: => String ) {
+private[ connector ] trait ExpectedFuture[ T ] {
+
+  protected def future: Future[ T ]
+
+  protected def cmd: String
 
   /** received an unexpected response */
-  private def onUnexpected( key: Option[ String ], cmd: => String ): PartialFunction[ Any, Nothing ] = {
-    case _ => unexpected( key, cmd )
-  }
+  protected def onUnexpected: PartialFunction[ Any, Nothing ]
+
+  protected def onFailed( ex: Throwable ): Nothing
 
   /** execution failed with an exception */
-  private def onException( key: Option[ String ], cmd: => String ): PartialFunction[ Throwable, Nothing ] = {
-    case ex => failed( key, cmd, ex )
+  private def onException: PartialFunction[ Throwable, Nothing ] = {
+    case ex => onFailed( ex )
   }
 
   /** handles both expected and unexpected responses and failure recovery */
   def expects[ U ]( expected: PartialFunction[ T, U ] )( implicit context: ExecutionContext ): Future[ U ] = {
-    future map ( expected orElse onUnexpected( Some( key ), cmd ) ) recover onException( Some( key ), cmd )
+    future map ( expected orElse onUnexpected ) recover onException
+  }
+}
+
+private[ connector ] class ExpectedFutureWithoutKey[ T ]( protected val future: Future[ T ], protected val cmd: String ) extends ExpectedFuture[ T ] {
+
+  protected def onUnexpected: PartialFunction[ Any, Nothing ] = {
+    case _ => unexpected( None, cmd )
   }
 
-  def withParameter( param: String ) = withParameters( param )
+  protected def onFailed( ex: Throwable ): Nothing =
+    failed( None, cmd, ex )
 
-  def withParameters( params: String ) = new ExpectedFuture( future, key, s"$key $params" )
+  def withKey( key: String ): ExpectedFutureWithKey[ T ] = new ExpectedFutureWithKey[ T ]( future, cmd, key, s"$cmd $key" )
+
+  def withKeys( keys: Traversable[ String ] ): ExpectedFutureWithKey[ T ] = withKey( keys mkString " " )
+}
+
+private[ connector ] class ExpectedFutureWithKey[ T ]( protected val future: Future[ T ], protected val cmd: String, key: String, fullCommand: => String ) extends ExpectedFuture[ T ] {
+
+  protected def onUnexpected: PartialFunction[ Any, Nothing ] = {
+    case _ => unexpected( Some( key ), cmd )
+  }
+
+  protected def onFailed( ex: Throwable ): Nothing =
+    failed( Some( key ), cmd, ex )
+
+  def andParameter( param: => Any ): ExpectedFutureWithKey[ T ] = andParameters( param.toString )
+
+  def andParameters( params: Traversable[ Any ] ): ExpectedFutureWithKey[ T ] = andParameters( params mkString " " )
+
+  def andParameters( params: => String ): ExpectedFutureWithKey[ T ] = new ExpectedFutureWithKey( future, cmd, key, s"$fullCommand $params" )
+
+  def asCommand( commandOverride: => String ) = new ExpectedFutureWithKey( future, cmd, key, s"$cmd $commandOverride" )
 }
 
 /**
   * Constructs expected future from provided parameters, this serves as syntax sugar
   */
-private[ connector ] class ExpectedFutureBuilder[ T ]( future: Future[ T ] ) {
+private[ connector ] class ExpectedFutureBuilder[ T ]( val future: Future[ T ] ) extends AnyVal {
 
-  def executing( key: String ): ExpectedFuture[ T ] = new ExpectedFuture[ T ]( future, key, key )
+  def executing( cmd: String ): ExpectedFutureWithoutKey[ T ] = new ExpectedFutureWithoutKey[ T ]( future, cmd )
 }

--- a/src/main/scala/play/api/cache/redis/connector/ExpectedFuture.scala
+++ b/src/main/scala/play/api/cache/redis/connector/ExpectedFuture.scala
@@ -23,6 +23,7 @@ private[ connector ] trait ExpectedFuture[ T ] {
 
   /** execution failed with an exception */
   private def onException: PartialFunction[ Throwable, Nothing ] = {
+    case ex: RedisException => throw ex
     case ex => onFailed( ex )
   }
 

--- a/src/main/scala/play/api/cache/redis/connector/RedisConnectorImpl.scala
+++ b/src/main/scala/play/api/cache/redis/connector/RedisConnectorImpl.scala
@@ -45,7 +45,7 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
   )
 
   def get[ T: ClassTag ]( key: String ): Future[ Option[ T ] ] =
-    redis.get[ String ]( key ) executing "GET" withParameter key expects {
+    redis.get[ String ]( key ) executing "GET" withKey key expects {
       case Some( response: String ) =>
         log.trace( s"Hit on key '$key'." )
         Some( decode[ T ]( key, response ) )
@@ -55,7 +55,7 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     }
 
   def mGet[ T: ClassTag ]( keys: String* ): Future[ Seq[ Option[ T ] ] ] =
-    redis.mget[ String ]( keys: _* ) executing "MGET" withParameters keys.mkString( " " ) expects {
+    redis.mget[ String ]( keys: _* ) executing "MGET" withKeys keys expects {
       // list is always returned
       case list => keys.zip( list ).map {
         case (key, Some( response: String )) =>
@@ -90,19 +90,19 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
 
   /** temporally stores already encoded value into the storage */
   private def setTemporally( key: String, value: String, expiration: Duration ): Future[ Unit ] =
-    redis.setex( key, expiration.toSeconds, value ) executing "SETEX" withParameters s"$key $value $expiration" expects {
+    redis.setex( key, expiration.toSeconds, value ) executing "SETEX" withKey key andParameters s"$value $expiration" expects {
       case _ => log.debug( s"Set on key '$key' on $expiration seconds." )
     }
 
   /** eternally stores already encoded value into the storage */
   private def setEternally( key: String, value: String ): Future[ Unit ] =
-    redis.set( key, value ) executing "SET" withParameters s"$key $value" expects {
+    redis.set( key, value ) executing "SET" withKey key andParameter value expects {
       case true => log.debug( s"Set on key '$key' for infinite seconds." )
       case false => log.warn( s"Set on key '$key' failed. Condition was not met." )
     }
 
   def setIfNotExists( key: String, value: Any ): Future[ Boolean ] =
-    encode( key, value ).flatMap( redis.setnx( key, _ ) ) executing "SETNX" withParameters s"$key encoded($value)" expects {
+    encode( key, value ).flatMap( redis.setnx( key, _ ) ) executing "SETNX" withKey key andParameter s"encoded($value)" expects {
       case false => log.debug( s"Set if not exists on key '$key' ignored. Value already exists." ); false
       case true => log.debug( s"Set if not exists on key '$key' succeeded." ); true
     }
@@ -124,25 +124,25 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
 
   /** eternally stores already encoded values into the storage */
   private def mSetEternally( keyValues: (String, String)* ): Future[ Unit ] =
-    redis.mset( keyValues.toMap ) executing "MSET" withParameters keyValues.map( _.asString ).mkString( " " ) expects {
+    redis.mset( keyValues.toMap ) executing "MSET" withKeys keyValues.map( _._1 ) asCommand keyValues.map( _.asString ).mkString( " " ) expects {
       case _ => log.debug( s"Set on keys ${ keyValues.map( _.key ) } for infinite seconds." )
     }
 
   /** eternally stores already encoded values into the storage */
   private def mSetEternallyIfNotExist( keyValues: (String, String)* ): Future[ Boolean ] =
-    redis.msetnx( keyValues.toMap ) executing "MSETNX" withParameters keyValues.map( _.asString ).mkString( " " ) expects {
+    redis.msetnx( keyValues.toMap ) executing "MSETNX" withKeys keyValues.map( _._1 ) asCommand keyValues.map( _.asString ).mkString( " " ) expects {
       case true => log.debug( s"Set if not exists on keys ${ keyValues.map( _.key ) mkString " " } succeeded." ); true
       case false => log.debug( s"Set if not exists on keys ${ keyValues.map( _.key ) mkString " " } ignored. Some value already exists." ); false
     }
 
   def expire( key: String, expiration: Duration ): Future[ Unit ] =
-    redis.expire( key, expiration.toSeconds.toInt ) executing "EXPIRE" withParameters s"$key, $expiration" expects {
+    redis.expire( key, expiration.toSeconds.toInt ) executing "EXPIRE" withKey key andParameter s"$expiration" expects {
       case true => log.debug( s"Expiration set on key '$key'." ) // expiration was set
       case false => log.debug( s"Expiration set on key '$key' failed. Key does not exist." ) // Nothing was removed
     }
 
   def matching( pattern: String ): Future[ Seq[ String ] ] =
-    redis.keys( pattern ) executing "KEYS" withParameter pattern expects {
+    redis.keys( pattern ) executing "KEYS" withKey pattern expects {
       case keys =>
         log.debug( s"KEYS on '$pattern' responded '${ keys.mkString( ", " ) }'." )
         keys
@@ -154,14 +154,14 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     }
 
   def exists( key: String ): Future[ Boolean ] =
-    redis.exists( key ) executing "EXISTS" withParameter key expects {
+    redis.exists( key ) executing "EXISTS" withKey key expects {
       case true => log.debug( s"Key '$key' exists." ); true
       case false => log.debug( s"Key '$key' doesn't exist." ); false
     }
 
   def remove( keys: String* ): Future[ Unit ] =
     if ( keys.nonEmpty ) // if any key to remove do it
-      redis.del( keys: _* ) executing "DEL" withParameters keys.mkString( " " ) expects {
+      redis.del( keys: _* ) executing "DEL" withKeys keys expects {
         // Nothing was removed
         case 0L => log.debug( s"Remove on keys ${ keys.mkString( "'", ",", "'" ) } succeeded but nothing was removed." )
         // Some entries were removed
@@ -175,17 +175,17 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     }
 
   def increment( key: String, by: Long ): Future[ Long ] =
-    redis.incrby( key, by ) executing "INCRBY" withParameters s"$key, $by" expects {
+    redis.incrby( key, by ) executing "INCRBY" withKey key andParameter s"$by" expects {
       case value => log.debug( s"The value at key '$key' was incremented by $by to $value." ); value
     }
 
   def append( key: String, value: String ): Future[ Long ] =
-    redis.append( key, value ) executing "APPEND" withParameters s"$key $value" expects {
+    redis.append( key, value ) executing "APPEND" withKey key andParameter value expects {
       case length => log.debug( s"The value was appended to key '$key'." ); length
     }
 
   def listPrepend( key: String, values: Any* ): Future[ Long ] =
-    Future.sequence( values.map( encode( key, _ ) ) ).flatMap( redis.lpush( key, _: _* ) ) executing "LPUSH" withParameters s"$key ${ values mkString " " }" expects {
+    Future.sequence( values.map( encode( key, _ ) ) ).flatMap( redis.lpush( key, _: _* ) ) executing "LPUSH" withKey key andParameters values expects {
       case length => log.debug( s"The $length values was prepended to key '$key'." ); length
     } recover {
       case ExecutionFailedException( _, _, ex ) if ex.getMessage startsWith "WRONGTYPE" =>
@@ -194,17 +194,17 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     }
 
   def listAppend( key: String, values: Any* ) =
-    Future.sequence( values.map( encode( key, _ ) ) ).flatMap( redis.rpush( key, _: _* ) ) executing "RPUSH" withParameters s"$key ${ values mkString " " }" expects {
+    Future.sequence( values.map( encode( key, _ ) ) ).flatMap( redis.rpush( key, _: _* ) ) executing "RPUSH" withKey key andParameters values expects {
       case length => log.debug( s"The $length values was appended to key '$key'." ); length
     }
 
   def listSize( key: String ) =
-    redis.llen( key ) executing "LLEN" withParameter key expects {
+    redis.llen( key ) executing "LLEN" withKey key expects {
       case length => log.debug( s"The collection at '$key' has $length items." ); length
     }
 
   def listSetAt( key: String, position: Int, value: Any ) =
-    encode( key, value ).flatMap( redis.lset( key, position, _ ) ) executing "LSET" withParameters s"$key $value" expects {
+    encode( key, value ).flatMap( redis.lset( key, position, _ ) ) executing "LSET" withKey key andParameter value expects {
       case _ => log.debug( s"Updated value at $position in '$key' to $value." )
     } recover {
       case ExecutionFailedException( _, _, actors.ReplyErrorException( "ERR index out of range" ) ) =>
@@ -213,7 +213,7 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     }
 
   def listHeadPop[ T: ClassTag ]( key: String ) =
-    redis.lpop[ String ]( key ) executing "LPOP" withParameter key expects {
+    redis.lpop[ String ]( key ) executing "LPOP" withKey key expects {
       case Some( encoded ) =>
         log.trace( s"Hit on head in key '$key'." )
         Some( decode[ T ]( key, encoded ) )
@@ -223,26 +223,26 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     }
 
   def listSlice[ T: ClassTag ]( key: String, start: Int, end: Int ) =
-    redis.lrange[ String ]( key, start, end ) executing "LRANGE" withParameters s"$start $end" expects {
+    redis.lrange[ String ]( key, start, end ) executing "LRANGE" withKey key andParameters s"$start $end" expects {
       case values =>
         log.debug( s"The range on '$key' from $start to $end included returned ${ values.size } values." )
         values.map( decode[ T ]( key, _ ) )
     }
 
   def listRemove( key: String, value: Any, count: Int ) =
-    encode( key, value ).flatMap( redis.lrem( key, count, _ ) ) executing "LREM" withParameter s"$key $value $count" expects {
+    encode( key, value ).flatMap( redis.lrem( key, count, _ ) ) executing "LREM" withKey key andParameters s"$value $count" expects {
       case removed => log.debug( s"Removed $removed occurrences of $value in '$key'." ); removed
     }
 
   def listTrim( key: String, start: Int, end: Int ) =
-    redis.ltrim( key, start, end ) executing "LTRIM" withParameter s"$key $start $end" expects {
+    redis.ltrim( key, start, end ) executing "LTRIM" withKey key andParameter s"$start $end" expects {
       case _ => log.debug( s"Trimmed collection at '$key' to $start:$end " )
     }
 
   def listInsert( key: String, pivot: Any, value: Any ) = for {
     pivot <- encode( key, pivot )
     value <- encode( key, value )
-    result <- redis.linsert( key, api.BEFORE, pivot, value ) executing "LINSERT" withParameter s"$key $pivot $value" expects {
+    result <- redis.linsert( key, api.BEFORE, pivot, value ) executing "LINSERT" withKey key andParameter s"$pivot $value" expects {
       case -1L | 0L => log.debug( s"Insert into the list at '$key' failed. Pivot not found." ); None
       case length => log.debug( s"Inserted $value into the list at '$key'. New size is $length." ); Some( length )
     } recover {
@@ -255,7 +255,7 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
   def setAdd( key: String, values: Any* ) = {
     // encodes the value
     def toEncoded( value: Any ) = encode( key, value )
-    Future.sequence( values map toEncoded ).flatMap( redis.sadd( key, _: _* ) ) executing "SADD" withParameters s"$key ${ values mkString " " }" expects {
+    Future.sequence( values map toEncoded ).flatMap( redis.sadd( key, _: _* ) ) executing "SADD" withKey key andParameters values expects {
       case inserted => log.debug( s"Inserted $inserted elements into the set at '$key'." ); inserted
     } recover {
       case ExecutionFailedException( _, _, ex ) if ex.getMessage startsWith "WRONGTYPE" =>
@@ -265,19 +265,19 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
   }
 
   def setSize( key: String ) =
-    redis.scard( key ) executing "SCARD" withParameter key expects {
+    redis.scard( key ) executing "SCARD" withKey key expects {
       case length => log.debug( s"The collection at '$key' has $length items." ); length
     }
 
   def setMembers[ T: ClassTag ]( key: String ) =
-    redis.smembers[ String ]( key ) executing "SMEMBERS" withParameter key expects {
+    redis.smembers[ String ]( key ) executing "SMEMBERS" withKey key expects {
       case items =>
         log.debug( s"Returned ${ items.size } items from the collection at '$key'." )
         items.map( decode[ T ]( key, _ ) ).toSet
     }
 
   def setIsMember( key: String, value: Any ) =
-    encode( key, value ).flatMap( redis.sismember( key, _ ) ) executing "SISMEMBER" withParameters s"$key $value" expects {
+    encode( key, value ).flatMap( redis.sismember( key, _ ) ) executing "SISMEMBER" withKey key andParameter value expects {
       case true => log.debug( s"Item $value exists in the collection at '$key'." ); true
       case false => log.debug( s"Item $value does not exist in the collection at '$key'" ); false
     }
@@ -286,46 +286,46 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     // encodes the value
     def toEncoded( value: Any ) = encode( key, value )
 
-    Future.sequence( values map toEncoded ).flatMap( redis.srem( key, _: _* ) ) executing "SREM" withParameters s"$key ${ values mkString " " }" expects {
+    Future.sequence( values map toEncoded ).flatMap( redis.srem( key, _: _* ) ) executing "SREM" withKey key andParameters values expects {
       case removed => log.debug( s"Removed $removed elements from the collection at '$key'." ); removed
     }
   }
 
   def hashRemove( key: String, fields: String* ) =
-    redis.hdel( key, fields: _* ) executing "HDEL" withParameters s"$key ${ fields mkString " " }" expects {
+    redis.hdel( key, fields: _* ) executing "HDEL" withKey key andParameters fields expects {
       case removed => log.debug( s"Removed $removed elements from the collection at '$key'." ); removed
     }
 
   def hashExists( key: String, field: String ) =
-    redis.hexists( key, field ) executing "HEXISTS" withParameters s"$key $field" expects {
+    redis.hexists( key, field ) executing "HEXISTS" withKey key andParameter field expects {
       case true => log.debug( s"Item $field exists in the collection at '$key'." ); true
       case false => log.debug( s"Item $field does not exist in the collection at '$key'" ); false
     }
 
   def hashGet[ T: ClassTag ]( key: String, field: String ) =
-    redis.hget[ String ]( key, field ) executing "HGET" withParameters s"$key $field" expects {
+    redis.hget[ String ]( key, field ) executing "HGET" withKey key andParameter field expects {
       case Some( encoded ) => log.debug( s"Item $field exists in the collection at '$key'." ); Some( decode[ T ]( key, encoded ) )
       case None => log.debug( s"Item $field is not in the collection at '$key'." ); None
     }
 
   def hashGetAll[ T: ClassTag ]( key: String ) =
-    redis.hgetall[ String ]( key ) executing "HGETALL" withParameter key expects {
+    redis.hgetall[ String ]( key ) executing "HGETALL" withKey key expects {
       case empty if empty.isEmpty => log.debug( s"Collection at '$key' is empty." ); Map.empty
       case encoded => log.debug( s"Collection at '$key' has ${ encoded.size } items." ); encoded.mapValues( decode[ T ]( key, _ ) )
     }
 
   def hashSize( key: String ) =
-    redis.hlen( key ) executing "HLEN" withParameter key expects {
+    redis.hlen( key ) executing "HLEN" withKey key expects {
       case length => log.debug( s"The collection at '$key' has $length items." ); length
     }
 
   def hashKeys( key: String ) =
-    redis.hkeys( key ) executing "HKEYS" withParameter key expects {
+    redis.hkeys( key ) executing "HKEYS" withKey key expects {
       case keys => log.debug( s"The collection at '$key' defines: ${ keys mkString " " }." ); keys.toSet
     }
 
   def hashSet( key: String, field: String, value: Any ) =
-    encode( key, value ).flatMap( redis.hset( key, field, _ ) ) executing "HSET" withParameters s"$key $field $value" expects {
+    encode( key, value ).flatMap( redis.hset( key, field, _ ) ) executing "HSET" withKey key andParameters s"$field $value" expects {
       case true => log.debug( s"Item $field in the collection at '$key' was inserted." ); true
       case false => log.debug( s"Item $field in the collection at '$key' was updated." ); false
     } recover {
@@ -335,7 +335,7 @@ private[ connector ] class RedisConnectorImpl @Inject()( serializer: AkkaSeriali
     }
 
   def hashValues[ T: ClassTag ]( key: String ) =
-    redis.hvals[ String ]( key ) executing "HVALS" withParameter key expects {
+    redis.hvals[ String ]( key ) executing "HVALS" withKey key expects {
       case values => log.debug( s"The collection at '$key' contains ${ values.size } values." ); values.map( decode[ T ]( key, _ ) ).toSet
     }
 

